sw/device/lib/dif/dif_uart.h

```diff
@@ -0,0 +1,280 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef _SW_DEVICE_LIB_DIF_UART_H_
+#define _SW_DEVICE_LIB_DIF_UART_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/dif/register.h"
+
+/**
+ * UART interrupt configuration
+ *
+ * Enumeration used to enable, disable, test and query the UART interrupts
+ */
+typedef enum dif_uart_interrupt {
+  dif_uart_tx_watermark =
+      0, /**< TX FIFO dipped below watermark TODO (pending HW changes) */
+  dif_uart_rx_watermark,  /**< RX FIFO reached high watermark */
+  dif_uart_tx_overflow,   /**< TX FIFO overflow */
+  dif_uart_rx_overflow,   /**< RX FIFO overflow */
+  dif_uart_rx_frame_err,  /**< RX framing error */
+  dif_uart_rx_break_err,  /**< RX break condition */
+  dif_uart_rx_timeout,    /**< RX FIFO timeout (not empty in a set time) */
+  dif_uart_rx_parity_err, /**< RX parity error detection */
+} dif_uart_interrupt_t;
+
+/**
+ * TX and RX FIFO depth configuration
+ *
+ * Enumeration used to set the upper limit of bytes to queue.
+ */
+typedef enum dif_uart_watermark {
+  dif_uart_watermark_byte_1 = 0, /**< 1 byte */
+  dif_uart_watermark_byte_4,     /**< 4 bytes */
+  dif_uart_watermark_byte_8,     /**< 8 bytes */
+  dif_uart_watermark_byte_16,    /**< 16 bytes */
+  dif_uart_watermark_byte_30,    /**< 30 bytes */
+} dif_uart_watermark_t;
```

Comment:
I think my comment here got splatted due to a force-push. These enumerators must be in `kConst` form, per the part of the Google C++ style guide we inherit. Our additions to the style guide for C notwithstanding; enumerators follow the rules for constants, not enum names. We already use the `kConst` form throughout the SW tree for this purpose, and I might file an update to the style guide to make this a bit more clear.

Reply:
I have change them up to have `kConst` form, but honestly I think for c benefit it's better to have them `MACRO_FORMAT`, which is allowed by the coding style guide 

>  Preferably, the individual enumerators should be named like constants. However, it is also acceptable to name them like macros.

Then we would avoid weird mixture like this:
```
...
  if (config->parity_enable == kDifUartEnable) {
    reg |= (1u << UART_CTRL_PARITY_EN);
  }
  if (config->parity == kDifUartOdd) {
    reg |= (1u << UART_CTRL_PARITY_ODD);
  ...
```

Reply:
Google C++ only allows that for compatibility reasons; all-caps should be reserved for true macros, to distinguish them from constants.

It's not a big deal, but I'm going to write `kConst` until I die, and we already use it exclusively throughout the SW tree (outside of fairly old stuff). Consistency matters.